File: crates/gravity/src/direct.rs
use rayon::prelude::*;

use crate::kernel::{kernel_accel_factor, kernel_potential_per_unit_mass, KernelKind};

// Tiny additive term to avoid division by zero in 1/sqrt(r2).
const R2_TINY: f64 = f64::MIN_POSITIVE;

// ===========================================================================
// Accumulator trait — allows a single generic loop for both accel & potential
// ===========================================================================

/// Abstracts over the output type so the direct-sum inner loops can be shared
/// between acceleration (`[f64; 3]`) and potential (`f64`) computations.
trait DirectAccumulator: Default + Clone + Send + Sync {
    type Output: Default + Send;

    /// Get a mutable reference to the accumulated value for particle `i`.
    fn init() -> Self;

    /// Accumulate the interaction from source `j` onto target `i`.
    ///
    /// `dx, dy, dz` = position difference (source - target).
    /// `m` = source mass.
    /// `invr` = 1/sqrt(r² + ε).
    fn add_newton(&mut self, dx: f64, dy: f64, dz: f64, m: f64, invr: f64);

    /// Accumulate a softened interaction.
    fn add_softened(&mut self, dx: f64, dy: f64, dz: f64, m: f64, r: f64, h: f64, kernel: KernelKind);

    /// Return the accumulated value, consuming self.
    fn finish(self) -> Self::Output;
}

/// Accumulator for gravitational acceleration `[ax, ay, az]`.
#[derive(Default, Clone)]
struct AccelAccumulator {
    ax: f64,
    ay: f64,
    az: f64,
}

impl DirectAccumulator for AccelAccumulator {
    type Output = [f64; 3];

    #[inline]
    fn init() -> Self {
        Self::default()
    }

    #[inline]
    fn add_newton(&mut self, dx: f64, dy: f64, dz: f64, m: f64, invr: f64) {
        // g = m / r^3  ->  g = m * invr * invr^2 = m * invr^3
        let invr2 = invr * invr;
        let invr3 = invr2 * invr;
        let g = m * invr3;
        self.ax += dx * g;
        self.ay += dy * g;
        self.az += dz * g;
    }

    #[inline]
    fn add_softened(&mut self, dx: f64, dy: f64, dz: f64, m: f64, r: f64, h: f64, kernel: KernelKind) {
        let g = m * kernel_accel_factor(kernel, r, h);
        self.ax += dx * g;
        self.ay += dy * g;
        self.az += dz * g;
    }

    #[inline]
    fn finish(self) -> Self::Output {
        [self.ax, self.ay, self.az]
    }
}

/// Accumulator for gravitational potential `phi`.
#[derive(Default, Clone)]
struct PotAccumulator {
    phi: f64,
}

impl DirectAccumulator for PotAccumulator {
    type Output = f64;

    #[inline]
    fn init() -> Self {
        Self::default()
    }

    #[inline]
    fn add_newton(&mut self, _dx: f64, _dy: f64, _dz: f64, m: f64, invr: f64) {
        self.phi += -m * invr;
    }

    #[inline]
    fn add_softened(&mut self, _dx: f64, _dy: f64, _dz: f64, m: f64, r: f64, h: f64, kernel: KernelKind) {
        self.phi += m * kernel_potential_per_unit_mass(kernel, r, h);
    }

    #[inline]
    fn finish(self) -> Self::Output {
        self.phi
    }
}

// ===========================================================================
// Generic direct-sum implementations
// ===========================================================================

/// Self-gravity: compute quantity on each source particle.
///
/// For small N (< 512) uses a symmetric pairwise loop (each pair computed once,
/// both particles updated).  For larger N uses a parallelised per-particle loop.
fn direct_self_impl<A: DirectAccumulator>(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    kernel: Option<KernelKind>,
) -> Vec<A::Output> {
    let n = positions.len();
    if n == 0 {
        return Vec::new();
    }

    let masses_slice_owned;
    let masses_slice: &[f64] = if let Some(m) = masses {
        m
    } else {
        masses_slice_owned = vec![1.0; n];
        &masses_slice_owned
    };

    let has_softening = softenings.is_some() && kernel.is_some();
    let kernel_kind = kernel.unwrap_or(KernelKind::Plummer);
    let kernel_is_spline = kernel_kind == KernelKind::CubicSplineW2;

    if n < 512 {
        // Symmetric pairwise loop — each (i,j) pair computed once.
        let mut accums: Vec<A> = (0..n).map(|_| A::init()).collect();

        for i in 0..n {
            let pi = positions[i];
            let mi = masses_slice[i];
            let hi = softenings.map(|hs| hs[i]).unwrap_or(0.0);

            for j in (i + 1)..n {
                let pj = positions[j];
                let mj = masses_slice[j];

                let dx = pj[0] - pi[0];
                let dy = pj[1] - pi[1];
                let dz = pj[2] - pi[2];
                let r2 = dx.mul_add(dx, dy.mul_add(dy, dz * dz));

                if has_softening {
                    let hj = softenings.map(|hs| hs[j]).unwrap_or(0.0);
                    let h = hi.max(hj);
                    if h <= 0.0 || (kernel_is_spline && r2 >= h * h) {
                        let invr = 1.0 / (r2 + R2_TINY).sqrt();
                        accums[i].add_newton(dx, dy, dz, mj, invr);
                        accums[j].add_newton(-dx, -dy, -dz, mi, invr);
                    } else {
                        let r = (r2 + R2_TINY).sqrt();
                        accums[i].add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                        accums[j].add_softened(-dx, -dy, -dz, mi, r, h, kernel_kind);
                    }
                } else {
                    let invr = 1.0 / (r2 + R2_TINY).sqrt();
                    accums[i].add_newton(dx, dy, dz, mj, invr);
                    accums[j].add_newton(-dx, -dy, -dz, mi, invr);
                }
            }
        }

        accums.into_iter().map(|a| a.finish()).collect()
    } else {
        // Parallel per-particle — each particle sums over all others.
        (0..n)
            .into_par_iter()
            .map(|i| {
                let mut acc = A::init();
                let pi = positions[i];
                let hi = softenings.map(|hs| hs[i]).unwrap_or(0.0);

                for j in 0..n {
                    if j == i {
                        continue;
                    }
                    let pj = positions[j];
                    let mj = masses_slice[j];

                    let dx = pj[0] - pi[0];
                    let dy = pj[1] - pi[1];
                    let dz = pj[2] - pi[2];
                    let r2 = dx.mul_add(dx, dy.mul_add(dy, dz * dz));

                    if has_softening {
                        let hj = softenings.map(|hs| hs[j]).unwrap_or(0.0);
                        let h = hi.max(hj);
                        if h <= 0.0 || (kernel_is_spline && r2 >= h * h) {
                            let invr = 1.0 / (r2 + R2_TINY).sqrt();
                            acc.add_newton(dx, dy, dz, mj, invr);
                        } else {
                            let r = (r2 + R2_TINY).sqrt();
                            acc.add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                        }
                    } else {
                        let invr = 1.0 / (r2 + R2_TINY).sqrt();
                        acc.add_newton(dx, dy, dz, mj, invr);
                    }
                }

                acc.finish()
            })
            .collect()
    }
}

/// Evaluate gravity from source particles at arbitrary target positions.
fn direct_at_points_impl<A: DirectAccumulator>(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    targets: &[[f64; 3]],
    kernel: Option<KernelKind>,
) -> Vec<A::Output> {
    let n_src = positions.len();
    let n_tgt = targets.len();
    if n_tgt == 0 || n_src == 0 {
        return Vec::new();
    }

    let masses_slice_owned;
    let masses_slice: &[f64] = if let Some(m) = masses {
        m
    } else {
        masses_slice_owned = vec![1.0; n_src];
        &masses_slice_owned
    };

    let has_softening = softenings.is_some() && kernel.is_some();
    let kernel_kind = kernel.unwrap_or(KernelKind::Plummer);
    let kernel_is_spline = kernel_kind == KernelKind::CubicSplineW2;

    if n_tgt < 512 {
        targets
            .iter()
            .map(|tgt| {
                let mut acc = A::init();
                let tx = tgt[0];
                let ty = tgt[1];
                let tz = tgt[2];

                for j in 0..n_src {
                    let pj = positions[j];
                    let mj = masses_slice[j];

                    let dx = pj[0] - tx;
                    let dy = pj[1] - ty;
                    let dz = pj[2] - tz;
                    let r2 = dx.mul_add(dx, dy.mul_add(dy, dz * dz));

                    if has_softening {
                        let hj = softenings.map(|hs| hs[j]).unwrap_or(0.0);
                        let h = hj.max(0.0);
                        if h <= 0.0 || (kernel_is_spline && r2 >= h * h) {
                            let invr = 1.0 / (r2 + R2_TINY).sqrt();
                            acc.add_newton(dx, dy, dz, mj, invr);
                        } else {
                            let r = (r2 + R2_TINY).sqrt();
                            acc.add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                        }
                    } else {
                        let invr = 1.0 / (r2 + R2_TINY).sqrt();
                        acc.add_newton(dx, dy, dz, mj, invr);
                    }
                }

                acc.finish()
            })
            .collect()
    } else {
        targets
            .par_iter()
            .map(|tgt| {
                let mut acc = A::init();
                let tx = tgt[0];
                let ty = tgt[1];
                let tz = tgt[2];

                for j in 0..n_src {
                    let pj = positions[j];
                    let mj = masses_slice[j];

                    let dx = pj[0] - tx;
                    let dy = pj[1] - ty;
                    let dz = pj[2] - tz;
                    let r2 = dx.mul_add(dx, dy.mul_add(dy, dz * dz));

                    if has_softening {
                        let hj = softenings.map(|hs| hs[j]).unwrap_or(0.0);
                        let h = hj.max(0.0);
                        if h <= 0.0 || (kernel_is_spline && r2 >= h * h) {
                            let invr = 1.0 / (r2 + R2_TINY).sqrt();
                            acc.add_newton(dx, dy, dz, mj, invr);
                        } else {
                            let r = (r2 + R2_TINY).sqrt();
                            acc.add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                        }
                    } else {
                        let invr = 1.0 / (r2 + R2_TINY).sqrt();
                        acc.add_newton(dx, dy, dz, mj, invr);
                    }
                }

                acc.finish()
            })
            .collect()
    }
}

// ===========================================================================
// Public API — thin wrappers over the generic implementations
// ===========================================================================

/// Direct-sum O(N²) gravitational accelerations (Newtonian, no softening).
pub fn direct_accelerations(positions: &[[f64; 3]], masses: Option<&[f64]>) -> Vec<[f64; 3]> {
    direct_self_impl::<AccelAccumulator>(positions, masses, None, None)
}

/// Direct-sum O(N²) gravitational accelerations at arbitrary target points.
pub fn direct_accelerations_at_points(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    targets: &[[f64; 3]],
) -> Vec<[f64; 3]> {
    direct_at_points_impl::<AccelAccumulator>(positions, masses, None, targets, None)
}

/// Direct-sum O(N²) gravitational potentials (Newtonian, no softening).
pub fn direct_potentials(positions: &[[f64; 3]], masses: Option<&[f64]>) -> Vec<f64> {
    direct_self_impl::<PotAccumulator>(positions, masses, None, None)
}

/// Direct-sum O(N²) gravitational potentials at arbitrary target points.
pub fn direct_potentials_at_points(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    targets: &[[f64; 3]],
) -> Vec<f64> {
    direct_at_points_impl::<PotAccumulator>(positions, masses, None, targets, None)
}

/// Direct-sum O(N²) gravitational potentials with softening kernel.
pub fn direct_potentials_kernel(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    kernel: KernelKind,
) -> Vec<f64> {
    direct_self_impl::<PotAccumulator>(positions, masses, softenings, Some(kernel))
}

/// Direct-sum O(N²) gravitational accelerations with softening kernel.
pub fn direct_accelerations_kernel(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    kernel: KernelKind,
) -> Vec<[f64; 3]> {
    direct_self_impl::<AccelAccumulator>(positions, masses, softenings, Some(kernel))
}

/// Direct-sum softened gravitational potentials at arbitrary target points.
pub fn direct_potentials_kernel_at_points(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    targets: &[[f64; 3]],
    kernel: KernelKind,
) -> Vec<f64> {
    direct_at_points_impl::<PotAccumulator>(positions, masses, softenings, targets, Some(kernel))
}

/// Direct-sum softened gravitational accelerations at arbitrary target points.
pub fn direct_accelerations_kernel_at_points(
    positions: &[[f64; 3]],
    masses: Option<&[f64]>,
    softenings: Option<&[f64]>,
    targets: &[[f64; 3]],
    kernel: KernelKind,
) -> Vec<[f64; 3]> {
    direct_at_points_impl::<AccelAccumulator>(positions, masses, softenings, targets, Some(kernel))
}

File: cpp/gravity/direct.cpp
#include "gravity/direct.hpp"
#include "gravity/common.hpp"
#include <cmath>
#include <cstddef>
#include <algorithm>

namespace gravity {

// ===========================================================================
// Generic direct-sum implementations
// ===========================================================================

// Self-gravity: compute quantity on each source particle.
//
// For small N (< 512) uses a symmetric pairwise loop (each pair computed once,
// both particles updated). For larger N uses a parallelised per-particle loop.
template <typename A>
std::vector<typename A::output_type> direct_self_impl(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, std::optional<KernelKind> kernel) {
    const std::size_t n = positions.size();
    if (n == 0) {
        return {};
    }

    std::vector<double> ones;
    const double* masses_slice = masses;
    if (masses_slice == nullptr) {
        ones.assign(n, 1.0);
        masses_slice = ones.data();
    }

    const bool has_softening = (softenings != nullptr) && kernel.has_value();
    const KernelKind kernel_kind = kernel.value_or(KernelKind::Plummer);
    const bool kernel_is_spline = kernel_kind == KernelKind::CubicSplineW2;

    if (n < 512) {
        // Symmetric pairwise loop — each (i,j) pair computed once.
        std::vector<A> accums(n);

        for (std::size_t i = 0; i < n; ++i) {
            const Vec3& pi = positions[i];
            const double mi = masses_slice[i];
            const double hi = softenings ? softenings[i] : 0.0;

            for (std::size_t j = i + 1; j < n; ++j) {
                const Vec3& pj = positions[j];
                const double mj = masses_slice[j];

                const double dx = pj.x - pi.x;
                const double dy = pj.y - pi.y;
                const double dz = pj.z - pi.z;
                const double r2 = dx * dx + dy * dy + dz * dz;

                if (has_softening) {
                    const double hj = softenings ? softenings[j] : 0.0;
                    const double h = std::max(hi, hj);
                    if (h <= 0.0 || (kernel_is_spline && r2 >= h * h)) {
                        const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                        accums[i].add_newton(dx, dy, dz, mj, invr);
                        accums[j].add_newton(-dx, -dy, -dz, mi, invr);
                    } else {
                        const double r = std::sqrt(r2 + R2_TINY);
                        accums[i].add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                        accums[j].add_softened(-dx, -dy, -dz, mi, r, h, kernel_kind);
                    }
                } else {
                    const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                    accums[i].add_newton(dx, dy, dz, mj, invr);
                    accums[j].add_newton(-dx, -dy, -dz, mi, invr);
                }
            }
        }

        std::vector<typename A::output_type> out;
        out.reserve(n);
        for (const A& a : accums) {
            out.push_back(a.finish());
        }
        return out;
    } else {
        // Parallel per-particle — each particle sums over all others.
        std::vector<typename A::output_type> out(n);
#pragma omp parallel for if(n >= 512) schedule(static)
        for (std::ptrdiff_t i = 0; i < static_cast<std::ptrdiff_t>(n); ++i) {
            A acc;
            const Vec3& pi = positions[static_cast<std::size_t>(i)];
            const double hi = softenings ? softenings[static_cast<std::size_t>(i)] : 0.0;

            for (std::size_t j = 0; j < n; ++j) {
                if (j == static_cast<std::size_t>(i)) {
                    continue;
                }
                const Vec3& pj = positions[j];
                const double mj = masses_slice[j];

                const double dx = pj.x - pi.x;
                const double dy = pj.y - pi.y;
                const double dz = pj.z - pi.z;
                const double r2 = dx * dx + dy * dy + dz * dz;

                if (has_softening) {
                    const double hj = softenings ? softenings[j] : 0.0;
                    const double h = std::max(hi, hj);
                    if (h <= 0.0 || (kernel_is_spline && r2 >= h * h)) {
                        const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                        acc.add_newton(dx, dy, dz, mj, invr);
                    } else {
                        const double r = std::sqrt(r2 + R2_TINY);
                        acc.add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                    }
                } else {
                    const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                    acc.add_newton(dx, dy, dz, mj, invr);
                }
            }

            out[static_cast<std::size_t>(i)] = acc.finish();
        }
        return out;
    }
}

// Evaluate gravity from source particles at arbitrary target positions.
template <typename A>
std::vector<typename A::output_type> direct_at_points_impl(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, const std::vector<Vec3>& targets,
    std::optional<KernelKind> kernel) {
    const std::size_t n_src = positions.size();
    const std::size_t n_tgt = targets.size();
    if (n_tgt == 0 || n_src == 0) {
        return {};
    }

    std::vector<double> ones;
    const double* masses_slice = masses;
    if (masses_slice == nullptr) {
        ones.assign(n_src, 1.0);
        masses_slice = ones.data();
    }

    const bool has_softening = (softenings != nullptr) && kernel.has_value();
    const KernelKind kernel_kind = kernel.value_or(KernelKind::Plummer);
    const bool kernel_is_spline = kernel_kind == KernelKind::CubicSplineW2;

    std::vector<typename A::output_type> out(n_tgt);
#pragma omp parallel for if(n_tgt >= 512)
    for (std::ptrdiff_t t = 0; t < static_cast<std::ptrdiff_t>(n_tgt); ++t) {
        A acc;
        const Vec3& tgt = targets[static_cast<std::size_t>(t)];
        const double tx = tgt.x;
        const double ty = tgt.y;
        const double tz = tgt.z;

        for (std::size_t j = 0; j < n_src; ++j) {
            const Vec3& pj = positions[j];
            const double mj = masses_slice[j];

            const double dx = pj.x - tx;
            const double dy = pj.y - ty;
            const double dz = pj.z - tz;
            const double r2 = dx * dx + dy * dy + dz * dz;

            if (has_softening) {
                const double hj = softenings ? softenings[j] : 0.0;
                const double h = std::max(hj, 0.0);
                if (h <= 0.0 || (kernel_is_spline && r2 >= h * h)) {
                    const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                    acc.add_newton(dx, dy, dz, mj, invr);
                } else {
                    const double r = std::sqrt(r2 + R2_TINY);
                    acc.add_softened(dx, dy, dz, mj, r, h, kernel_kind);
                }
            } else {
                const double invr = 1.0 / std::sqrt(r2 + R2_TINY);
                acc.add_newton(dx, dy, dz, mj, invr);
            }
        }

        out[static_cast<std::size_t>(t)] = acc.finish();
    }
    return out;
}

// ===========================================================================
// Public API — thin wrappers over the generic implementations
// ===========================================================================

std::vector<Vec3> direct_accelerations(const std::vector<Vec3>& positions, const double* masses) {
    return direct_self_impl<AccelAccumulator>(positions, masses, nullptr, std::nullopt);
}

std::vector<Vec3> direct_accelerations_at_points(
    const std::vector<Vec3>& positions, const double* masses, const std::vector<Vec3>& targets) {
    return direct_at_points_impl<AccelAccumulator>(positions, masses, nullptr, targets, std::nullopt);
}

std::vector<double> direct_potentials(const std::vector<Vec3>& positions, const double* masses) {
    return direct_self_impl<PotAccumulator>(positions, masses, nullptr, std::nullopt);
}

std::vector<double> direct_potentials_at_points(
    const std::vector<Vec3>& positions, const double* masses, const std::vector<Vec3>& targets) {
    return direct_at_points_impl<PotAccumulator>(positions, masses, nullptr, targets, std::nullopt);
}

std::vector<double> direct_potentials_kernel(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, KernelKind kernel) {
    return direct_self_impl<PotAccumulator>(positions, masses, softenings, std::optional<KernelKind>(kernel));
}

std::vector<Vec3> direct_accelerations_kernel(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, KernelKind kernel) {
    return direct_self_impl<AccelAccumulator>(positions, masses, softenings, std::optional<KernelKind>(kernel));
}

std::vector<double> direct_potentials_kernel_at_points(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, const std::vector<Vec3>& targets, KernelKind kernel) {
    return direct_at_points_impl<PotAccumulator>(positions, masses, softenings, targets, std::optional<KernelKind>(kernel));
}

std::vector<Vec3> direct_accelerations_kernel_at_points(
    const std::vector<Vec3>& positions, const double* masses,
    const double* softenings, const std::vector<Vec3>& targets, KernelKind kernel) {
    return direct_at_points_impl<AccelAccumulator>(positions, masses, softenings, targets, std::optional<KernelKind>(kernel));
}

// ===========================================================================
// Explicit template instantiations
// ===========================================================================

template std::vector<Vec3> direct_self_impl<AccelAccumulator>(const std::vector<Vec3>&, const double*, const double*, std::optional<KernelKind>);
template std::vector<double> direct_self_impl<PotAccumulator>(const std::vector<Vec3>&, const double*, const double*, std::optional<KernelKind>);
template std::vector<Vec3> direct_at_points_impl<AccelAccumulator>(const std::vector<Vec3>&, const double*, const double*, const std::vector<Vec3>&, std::optional<KernelKind>);
template std::vector<double> direct_at_points_impl<PotAccumulator>(const std::vector<Vec3>&, const double*, const double*, const std::vector<Vec3>&, std::optional<KernelKind>);

} // namespace gravity
